kernel/syscall.c

```diff
@@ -8,17 +6,27 @@
 #include "com.h"
 #include "hpet.h"
 
+typedef void __attribute__((noreturn)) (*syscall_func) (void);
+
+/* Syscall functions must never return.  Therefore, here are some helpers for
+ * wrapping methods that always return a value (without switching threads or
+ * anything), and making them instead never return.
+ */
+
 #pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-parameter"
+#pragma GCC diagnostic ignored "-Wunused-variable"
 #pragma GCC diagnostic ignored "-Wint-conversion"
 
 #define TRAMPOLINE_NAME(n) syscall_trampoline_ ## n
 
 #define TRAMPOLINE_HELPER(n, x) \
-  static syscall_arg TRAMPOLINE_NAME(n) ( syscall_arg a, syscall_arg b ) { \
-    syscall_arg res = 0; \
-    x;\
-    return res; \
+  static void __attribute__((noreturn)) TRAMPOLINE_NAME(n) (void) { \
+    uint64_t a = running_tcb->saved_registers.rbx; \
+    uint64_t b = running_tcb->saved_registers.rcx; \
+    uint64_t res = running_tcb->saved_registers.rax; \
+    x; \
+    running_tcb->saved_registers.rax = res; \
```

Comment:
Perhaps `res` should be a pointer to `rax` instead or just do `running_tcp->saved_registers.rax = x` and have the NORET ones use the comma operator so they always return zero or something.

Alternatively perhaps the NORET ones should go away.

Reply:
Thinking about it more, even invalid syscalls return a status, so I think returning 0 makes sense.  I used the comma operator.